File: app/src/main/java/com/sysu/edu/view/GridDialog.java
package com.sysu.edu.view;

import android.content.DialogInterface;
import android.content.res.ColorStateList;
import android.view.View;
import android.widget.GridLayout;

import androidx.appcompat.content.res.AppCompatResources;
import androidx.fragment.app.FragmentActivity;

import com.google.android.material.bottomsheet.BottomSheetBehavior;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.button.MaterialButton;
import com.sysu.edu.R;
import com.sysu.edu.databinding.DialogGridBinding;
import com.sysu.edu.databinding.ItemButtonGridBinding;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.stream.IntStream;

public class GridDialog {
    
    final ArrayList<Integer> referenceIds = new ArrayList<>();
    private final DialogGridBinding menuBinding;
    private final BottomSheetDialog menuDialog;
    private final FragmentActivity activity;
    private int selected = -1;
    private boolean selectable = false;
    private boolean multipleSelectable = false;
    private int iconGravity;
    private int gravity;
    
    public GridDialog(FragmentActivity activity) {
        this.activity = activity;
        menuDialog = new BottomSheetDialog(activity);
        menuBinding = DialogGridBinding.inflate(activity.getLayoutInflater());
        menuDialog.setContentView(menuBinding.getRoot());
    }
    
    
    public void setColumn(int column) {
        ((GridLayout.LayoutParams) menuBinding.handler.getLayoutParams()).columnSpec = GridLayout.spec(GridLayout.UNDEFINED, column, GridLayout.FILL, 1.0f);
        menuBinding.grid.setColumnCount(column);
    }
    
    public void setIconGravity(int gravity) {
        iconGravity = gravity;
        referenceIds.forEach(id -> ((MaterialButton) menuBinding.grid.findViewById(id)).setIconGravity(gravity));
    }
    
    public void setGravity(int gravity) {
        this.gravity = gravity;
        referenceIds.forEach(id -> ((MaterialButton) menuBinding.grid.findViewById(id)).setGravity(gravity));
    }
    
    public <T> void loadMenu(List<T> menuTitle, List<Integer> menuIcon, List<? extends Consumer<MaterialButton>> menuAction, Class<T> type) {
        referenceIds.clear();
        IntStream.range(0, menuTitle.size()).forEach(i -> {
            MaterialButton menu = ItemButtonGridBinding.inflate(activity.getLayoutInflater(), menuBinding.grid, false).getRoot();
            if (type == Integer.class)
                menu.setText((Integer) menuTitle.get(i));
            else
                menu.setText((String) menuTitle.get(i));
            if (menuIcon.size() > i && menuIcon.get(i) != 0)
                menu.setIconResource(menuIcon.get(i));
            int id = View.generateViewId();
            referenceIds.add(id);
            menu.setId(id);
            menu.addOnCheckedChangeListener((_, isChecked) -> menu.setStrokeWidth((isChecked && (selectable || multipleSelectable)) ? 3 : 0));
            menu.setOnClickListener(_ -> {
                if (menuAction.size() > i && menuAction.get(i) != null)
                    menuAction.get(i).accept(menu);
                if (multipleSelectable)
                    toggleMenu(i);
                else if (selectable)
                    selectMenu(i);
            });
            if (iconGravity != 0)
                menu.setIconGravity(iconGravity);
            if (gravity != 0)
                menu.setGravity(gravity);
            menuBinding.grid.addView(menu);
        });
    }
    
    public void show() {
        menuDialog.show();
        menuDialog.getBehavior().setState(BottomSheetBehavior.STATE_EXPANDED);
    }
    
    public BottomSheetDialog getDialog() {
        return menuDialog;
    }
    
    public MaterialButton getMenu(int position) {
        return position >= referenceIds.size() || position < 0 ? null : menuBinding.grid.findViewById(referenceIds.get(position));
    }
    
    public void setSelectable(boolean selectable) {
        this.selectable = selectable;
    }
    
    public void setMultipleSelectable(boolean multipleSelectable) {
        this.multipleSelectable = multipleSelectable;
    }
    
    public void selectMenu(int position) {
        MaterialButton menu = getMenu(position);
        if (menu != null && position != selected) {
            menu.setChecked(true);
            if (selected >= 0) {
                MaterialButton selectedMenu = getMenu(selected);
                if (selectedMenu != null)
                    selectedMenu.setChecked(false);
            }
            selected = position;
        }
    }
    
    public void toggleMenu(int position) {
        MaterialButton menu = getMenu(position);
        if (menu != null)
            menu.setChecked(!menu.isChecked());
    }
    
    public void toggleMenu(int position, boolean toggle) {
        MaterialButton menu = getMenu(position);
        if (menu != null)
            menu.setChecked(toggle);
    }
    
    public void clickMenu(int position) {
        MaterialButton menu = getMenu(position);
        if (menu != null)
            menu.performClick();
    }


    /*public void multipleSelectMenu(int[] positions) {
        IntStream.of(positions).forEach(this::selectMenu);
    }*/
    
    public void setTogglable(int[] positions, boolean togglable) {
        IntStream.of(positions).forEach(i -> setTogglable(i, togglable));
    }
    
    public void setTogglable(int position, boolean togglable) {
        MaterialButton menu = getMenu(position);
        if (menu != null) {
            ColorStateList colorStateList = AppCompatResources.getColorStateList(activity, R.color.toggle);
            menu.setToggleCheckedStateOnClick(togglable);
            menu.setCheckable(togglable);
            menu.setIconTint(togglable ? colorStateList : null);
            menu.setTextColor(togglable ? colorStateList : null);
        }
    }
    
    public void setPositiveButton(CharSequence text, DialogInterface.OnClickListener action) {
        menuBinding.positive.setText(text);
        menuBinding.positive.setOnClickListener(_ -> action.onClick(menuDialog, DialogInterface.BUTTON_POSITIVE));
        menuBinding.buttonGroup.setVisibility(View.VISIBLE);
    }
    
    public void setPositiveButton(int text, DialogInterface.OnClickListener action) {
        menuBinding.positive.setText(text);
        menuBinding.positive.setOnClickListener(_ -> action.onClick(menuDialog, DialogInterface.BUTTON_POSITIVE));
        menuBinding.buttonGroup.setVisibility(View.VISIBLE);
    }
    
    public void setNegativeButton(CharSequence text, DialogInterface.OnClickListener action) {
        menuBinding.negative.setText(text);
        menuBinding.negative.setOnClickListener(_ -> action.onClick(menuDialog, DialogInterface.BUTTON_NEGATIVE));
        menuBinding.buttonGroup.setVisibility(View.VISIBLE);
    }
    
    public void setNegativeButton(int text, DialogInterface.OnClickListener action) {
        menuBinding.negative.setText(text);
        menuBinding.negative.setOnClickListener(_ -> action.onClick(menuDialog, DialogInterface.BUTTON_NEGATIVE));
        menuBinding.buttonGroup.setVisibility(View.VISIBLE);
    }
    
    public void setNeutralButton(CharSequence text, DialogInterface.OnClickListener action) {
        menuBinding.neutral.setText(text);
        menuBinding.neutral.setOnClickListener(_ -> action.onClick(menuDialog, DialogInterface.BUTTON_NEUTRAL));
        menuBinding.buttonGroup.setVisibility(View.VISIBLE);
    }
    
    public void setNeutralButton(int text, DialogInterface.OnClickListener action) {
        menuBinding.neutral.setText(text);
        menuBinding.neutral.setOnClickListener(_ -> action.onClick(menuDialog, DialogInterface.BUTTON_NEUTRAL));
        menuBinding.buttonGroup.setVisibility(View.VISIBLE);
    }
    
    public void dismiss() {
        menuDialog.dismiss();
    }
}

File: app/src/main/java/com/sysu/edu/view/GridDialog.kt
package com.sysu.edu.view

import android.content.DialogInterface
import android.view.View
import android.widget.GridLayout
import androidx.appcompat.content.res.AppCompatResources
import androidx.fragment.app.FragmentActivity
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.button.MaterialButton
import com.sysu.edu.R
import com.sysu.edu.databinding.DialogGridBinding
import com.sysu.edu.databinding.ItemButtonGridBinding
import java.util.function.Consumer

class GridDialog(private val activity: FragmentActivity) {
	val referenceIds: MutableList<Int?> = mutableListOf()
	private val menuBinding: DialogGridBinding = DialogGridBinding.inflate(activity.layoutInflater)
	val dialog: BottomSheetDialog = BottomSheetDialog(activity)
	private var selected = -1
	private var selectable = false
	private var multipleSelectable = false
	private var iconGravity = 0
	private var gravity = 0
	
	init {
		dialog.setContentView(menuBinding.root)
	}
	
	fun setColumn(column: Int) {
		(menuBinding.handler.layoutParams as GridLayout.LayoutParams).columnSpec = GridLayout.spec(GridLayout.UNDEFINED, column, GridLayout.FILL, 1.0f)
		menuBinding.grid.setColumnCount(column)
	}
	
	fun setIconGravity(gravity: Int) {
		iconGravity = gravity
		referenceIds.forEach { id: Int? -> (menuBinding.grid.findViewById<View?>(id!!) as MaterialButton).iconGravity = gravity }
	}
	
	fun setGravity(gravity: Int) {
		this.gravity = gravity
		referenceIds.forEach { id: Int? -> (menuBinding.grid.findViewById<View?>(id!!) as MaterialButton).setGravity(gravity) }
	}
	
	fun <T> loadMenu(menuTitle: MutableList<T?>,
	                 menuIcon: MutableList<Int?>,
	                 menuAction: MutableList<out Consumer<MaterialButton?>?>,
	                 type: Class<T?>?) {
		referenceIds.clear()
		menuTitle.forEachIndexed { i, v ->
			val menu = ItemButtonGridBinding.inflate(activity.layoutInflater, menuBinding.grid, false).root
			if (type == Int::class.java) menu.setText((v as Int?)!!)
			else menu.text = v as String?
			if (menuIcon.size > i && menuIcon[i] != 0) menu.setIconResource(menuIcon[i]!!)
			val id = View.generateViewId()
			referenceIds.add(id)
			menu.setId(id)
			menu.addOnCheckedChangeListener { _: MaterialButton?, isChecked: Boolean -> menu.strokeWidth = if (isChecked && (selectable || multipleSelectable)) 3 else 0 }
			menu.setOnClickListener {
				if (menuAction.size > i && menuAction[i] != null) menuAction[i]!!.accept(menu)
				if (multipleSelectable) toggleMenu(i)
				else if (selectable) selectMenu(i)
			}
			if (iconGravity != 0) menu.iconGravity = iconGravity
			if (gravity != 0) menu.setGravity(gravity)
			menuBinding.grid.addView(menu)
		}
	}
	
	fun show() {
		dialog.show()
		dialog.behavior.setState(BottomSheetBehavior.STATE_EXPANDED)
	}
	
	fun getMenu(position: Int): MaterialButton? {
		return if (position >= referenceIds.size || position < 0) null else menuBinding.grid.findViewById(referenceIds[position]!!)
	}
	
	fun setSelectable(selectable: Boolean) {
		this.selectable = selectable
	}
	
	fun setMultipleSelectable(multipleSelectable: Boolean) {
		this.multipleSelectable = multipleSelectable
	}
	
	fun selectMenu(position: Int) {
		val menu = getMenu(position)
		if (menu != null && position != selected) {
			menu.isChecked = true
			if (selected >= 0) getMenu(selected)?.isChecked = false
			selected = position
		}
	}
	
	fun toggleMenu(position: Int) {
		val menu = getMenu(position)
		if (menu != null) menu.isChecked = !menu.isChecked
	}
	
	fun toggleMenu(position: Int, toggle: Boolean) {
		getMenu(position)?.isChecked = toggle
	}
	
	fun clickMenu(position: Int) {
		getMenu(position)?.performClick()
	}
	
	/*public void multipleSelectMenu(int[] positions) {
        IntStream.of(positions).forEach(this::selectMenu);
    }*/
	fun setTogglable(positions: IntArray, togglable: Boolean) {
		positions.forEach { setTogglable(it, togglable) }
	}
	
	fun setTogglable(position: Int, togglable: Boolean) {
		getMenu(position)?.run {
			val colorStateList = AppCompatResources.getColorStateList(activity, R.color.toggle)
			isToggleCheckedStateOnClick = togglable
			isCheckable = togglable
			setIconTint(if (togglable) colorStateList else null)
			setTextColor(if (togglable) colorStateList else null)
		}
	}
	
	fun setPositiveButton(text: CharSequence?, action: DialogInterface.OnClickListener) {
		menuBinding.positive.text = text
		menuBinding.positive.setOnClickListener { action.onClick(this.dialog, DialogInterface.BUTTON_POSITIVE) }
		menuBinding.buttonGroup.setVisibility(View.VISIBLE)
	}
	
	fun setPositiveButton(text: Int, action: DialogInterface.OnClickListener) {
		menuBinding.positive.setText(text)
		menuBinding.positive.setOnClickListener { action.onClick(this.dialog, DialogInterface.BUTTON_POSITIVE) }
		menuBinding.buttonGroup.setVisibility(View.VISIBLE)
	}
	
	fun setNegativeButton(text: CharSequence?, action: DialogInterface.OnClickListener) {
		menuBinding.negative.text = text
		menuBinding.negative.setOnClickListener { action.onClick(this.dialog, DialogInterface.BUTTON_NEGATIVE) }
		menuBinding.buttonGroup.setVisibility(View.VISIBLE)
	}
	
	fun setNegativeButton(text: Int, action: DialogInterface.OnClickListener) {
		menuBinding.negative.setText(text)
		menuBinding.negative.setOnClickListener { action.onClick(this.dialog, DialogInterface.BUTTON_NEGATIVE) }
		menuBinding.buttonGroup.setVisibility(View.VISIBLE)
	}
	
	fun setNeutralButton(text: CharSequence?, action: DialogInterface.OnClickListener) {
		menuBinding.neutral.text = text
		menuBinding.neutral.setOnClickListener { action.onClick(this.dialog, DialogInterface.BUTTON_NEUTRAL) }
		menuBinding.buttonGroup.setVisibility(View.VISIBLE)
	}
	
	fun setNeutralButton(text: Int, action: DialogInterface.OnClickListener) {
		menuBinding.neutral.setText(text)
		menuBinding.neutral.setOnClickListener { action.onClick(this.dialog, DialogInterface.BUTTON_NEUTRAL) }
		menuBinding.buttonGroup.setVisibility(View.VISIBLE)
	}
	
	fun dismiss() {
		dialog.dismiss()
	}
}
